src/Microsoft.AspNet.OData.Shared/Builder/ODataInstanceAnnotationContainer.cs

```diff
@@ -0,0 +1,125 @@
+        /// </summary>
+        public ODataInstanceAnnotationContainer()
+        {
+            instanceAnnotations = new Dictionary<string, IDictionary<string, object>>();
+        }
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to the CLR type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        public void AddResourceAnnotation(string annotationName, object value)
+        {
+            SetInstanceAnnotation(string.Empty, annotationName, value);
+        }
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to a property
+        /// </summary>
+        /// <param name="propertyName">Name of the property</param>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        public void AddPropertyAnnotation(string propertyName, string annotationName, object value)
+        {
+            SetInstanceAnnotation(propertyName, annotationName, value);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from CLR Type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <returns>Get Annotation value for the given annotation</returns>
+        public object GetResourceAnnotation(string annotationName)
+        {
+            return GetInstanceAnnotation(string.Empty, annotationName);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from the Property
+        /// </summary>
+        /// <param name="propertyName">Name of the Property</param>
+        /// <param name="annotationName">Name of the Annotation</param>
+        /// <returns>Get Annotation value for the given annotation and property</returns>
+        public object GetPropertyAnnotation(string propertyName, string annotationName)
+        {
+            return GetInstanceAnnotation(propertyName, annotationName);
+        }
+
+        /// <summary>
+        /// Get All Annotations from CLR Type
+        /// </summary>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        public IDictionary<string, object> GetAllResourceAnnotations()
+        {
+            return GetAllInstanceAnnotations(string.Empty);
+        }
+
+        /// <summary>
+        /// Get all Annotation for a Property
+        /// </summary>
+        /// <param name="propertyName">Name of Property</param>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        public IDictionary<string, object> GetAllPropertyAnnotations(string propertyName)
+        {
+            return GetAllInstanceAnnotations(propertyName);
+        }
+
+        private void SetInstanceAnnotation(string propertyName, string annotationName, object value)
```

Comment:
>annotationName [](start = 71, length = 14)

shall we verify  the annotation name format?

Reply:
This is actually a good point.  ODL will validate at serialization, but it would be nice if our class caught this at set time.

Since "@" has special meaning, we should validate that the name not start with "@" and must contain at least one ".".

---
In reply to: [490459975](https://github.com/OData/WebApi/pull/2219#discussion_r490459975) [](ancestors = 490459975)